README.md
# Gamecord

Easy to make discord games for your bot by us!

# Installation

```bash
npm install gamecord
```

# Features
- Super easy to use 😀

# Quick Example

```js
// Install Gamecord
const GameCord = require('gamecord');

// Create Client
const { Client } = require('discord.js');
const client = new Client();

// Ready Event
client.on('ready', () => console.log('Bot is ready :D'));

// Message Event
client.on('message', message => {
    if(message.content == '!hangman'){

        new GameCord.HangmanGame(message)
        .setTitle('My Hangman')
        .setColor('#7298da')
        .setHint() // Only if you want a hint for your hangman!
        .setTime(20000) // Default is 30 secs
        .on('end', game => console.log(`${game.user.tag} ${game.win ? 'win' : 'lose'} the game!`))
        .on('start', game => console.log(`${game.message.author.tag} started a hangman game with word ${game.word}!`))
        .run() // Keep all your settings above and run it after all of your configuration!

    }else if(message.content == '!snake'){

        new GameCord.SnakeGame(message)
        .setTitle('My snake')
        .setColor('#7298da')
        .setMaxTime(60000) // Always better to set max time because the default one is just 5s
        .on('end', game => console.log(`${game.message.author.tag}'s snake game score was ${game.score}`)) // Start event also exists
        .run()

    }
});

// Login the bot
client.login(process.env.token);
```

# Quick Methods

- `Minesweeper({ width: number, height: number })`

Will return a grid of minesweeper using discord spoiler tag...

```js
message.channel.send(
    new Discord.MessageEmbed()
    .setTitle('My Minesweeper')
    .setColor('#7298da')
    .setDescription(GameCord.Minesweeper()) // By default minesweeper method takes 8x8 grid
    .setTimestamp()
)
```

# Samples
![1](/images/snake.png)

![2](/images/hangman.png)

![3](/images/mine.png)

# Authors
* **[1GPEX](https://github.com/1GPEX)** - *Original Idea & Make snake,connect4 ganes*
* **[Science Spot](https://github.com/Scientific-Guy)** - *Make a options & hangman,snake,minesweeper games* 

# Applicable Games
- Hangman
- Snake
- Minesweeper
- ConnectFour

### [Discord](https://discord.gg/hw7XPxz) - *Join discord server for help*


CHANGELOG.md
# Gamecord

not yet released but just created in advanced

src/Hangman.js
/**
 * Hangman Game
 */

const { EventEmitter } = require('events');
const { random, words } = require('./utils/index');

class HangmanGame{

    /**
     * Make your hangman game...
     * @param {any} message Client Message
     * @param {object} [options={}] options 
     * @example const GameCord = require('gamecord');
     * new GameCord.HangmanGame(message)
        .setTitle('My Hangman')
        .setColor('#7298da')
        .setTime(20000) // Default is 30 secs
        .setHint() // Only if you want a hint for your hangman!
        .on('end', game => console.log(`${game.user.tag} ${game.win ? 'win' : 'lose'} the game!`))
        .on('start', game => console.log(`${game.message.author.tag} started a hangman game with word ${game.word}!`))
        .run() // Keep all your settings above and run it after all of your configuration!
     */
    constructor(message, options={}){
        if(!message) throw new Error('message param is required');

        this.message = message;
        this.event = new EventEmitter();
        this.mistakes = 0;
        this.word = null;
        this.wordArray = [];
        this.embed = null;
        this.guessed = [];

        this.options = {
            title: 'Hangman',
            color: 'RANDOM',
            gameOverTitle: 'Game Over',
            hint: false,
            time: 300000,
            words,
            ...options
        };
    };

    /**
     * Get the hangman description
     * @example HangmanGame.description
     */
    get description(){
        return "```"
            + (this.mistakes > 0 ? "|‾‾‾‾‾‾|   \n|     " : " ")
            + (this.mistakes > 1 ? " |" : " ")
            + "   \n|     "
            + (this.mistakes > 2 ? " O" : " ")
            + "   \n|     "
            + (this.mistakes > 3 ? "/|\\" : " ")
            + "   \n|     "
            + (this.mistakes > 4 ? "/ \\" : " ")
            + "   \n|     \n|__________"
            + (this.options.hint ? `\n\n${this.hint.join(' ')}` : '')
            + "```\n"
    };

    /**
     * Hint gen
     */
    get hint(){
        return this.wordArray.map(x => this.guessed.includes(x) ? x : '_')
    }

    /**
     * Run the hangman game
     * @example HangmanGame.run()
     */
    run(){
        this.word = this.options.words[Math.floor(Math.random() * this.options.words.length)].toUpperCase();
        this.wordArray = this.word.split('');
        if(this.options.hint) for(let i = 0; i < 2; i++) this.guessed.push(random(this.wordArray));
        this.event.emit('start', this);

        this.message.channel.send({
            embed: {
                title: this.options.title,
                color: this.options.color,
                description: this.description,
                timestamp: Date.now(),
                footer: {
                    text: 'Type your guess below!'
                }
            }
        }).then(message => {
            this.embed = message;
            let win = true;
            this.message.channel.awaitMessages(this.next, { max: 1, time: this.options.time, errors: ['time'] })
            .then(() => this.end(win)).catch(() => this.end(0));
        });
    };

    /**
     * Next event
     * @param {any} m Message object 
     */
    next(m){
        if(m.author.id != this.message.author.id) return false;
        this.guessed.push(m.content.toUpperCase());
        if(!this.hint.includes('_')) return true;
        if(!this.wordArray.includes(m.content.toUpperCase())) this.mistakes++;
        if(this.mistakes == 6){ win = false; return true; };
        this.edit();
    };

    /**
     * Edits the embed
     */
    edit(){
        this.embed.edit({
            embed: {
                title: this.options.title,
                color: this.options.color,
                description: this.description,
                timestamp: Date.now(),
                footer: {
                    text: 'Type your guess below!'
                }
            }
        });
    };

    /**
     * Function to end the game
     * @param {boolean} win A boolean deciding if the user has won or not
     * @example HangmanGame.end(true)
     */
    end(win){
        this.embed.edit({
            embed: {
                title: this.options.gameOverTitle,
                color: this.options.color,
                description: `**${win ? 'You won!' : 'You lose!'}**\n**The Word:** ${this.word}`,
                timestamp: Date.now()
            }
        });

        this.event.emit('end', { win, ...this });
    };

    /**
     * Set title
     * @param {string} title Title to set
     */
    setTitle(title){
        this.options.title = title;
        return this;
    };

    /**
     * Set color
     * @param {string} color Custom color to set
     */
    setColor(color){
        this.options.color = color;
        return this;
    };

    /**
     * Set error time
     * @param {number} time Number to set for message collector
     */
    setTime(time){
        this.options.time = time;
        return this;
    };

    /**
     * Set game over title
     * @param {string} title Title to set for game over embed
     */
    setGameOverTitle(title){
        this.options.gameOverTitle = title;
        return this;
    };

    /**
     * Set your own custom words to ask as quiz
     * @param {any[]} words Words to set to quiz
     */
    setWords(words){
        if(!Array.isArray(words)) throw new Error('invalid set of words');
        this.options.words = words;
        return this;
    };

    /**
     * Push your custom words to the existence list
     * @param {any[]} words Words to push
     */
    pushWords(words){
        if(!Array.isArray(words)) throw new Error('invalid set of words');
        this.options.words = this.options.words.concat(words);
        return this;
    };

    /**
     * Will set hint settings to true
     */
    setHint(){
        this.options.hint = true;
        return this;
    };

    /**
     * Set function for the events
     * @param {string} event Evebt name
     * @param {any} callback Event Function
     */
    on(event, callback){
        this.event.on(event, callback);
        return this;
    };

};

module.exports = HangmanGame